File: lib/vpago/acleda_mobile/callback_validator.php
<?php
  $data = array(
    "TransactionId" => "REF0361472663",
    "PaymentTokenId" => "16de81d4-b5ef-ef59-16de-81d4b5efef59",
    "TxnAmount" => "30",
    "SenderName" => "Test User"
  );

  // will be provided
  $secret_key = 'VTenhSecret';

  // signed request fields
  $hash_data = $data['TransactionId'] . ' ' . $data['PaymentTokenId'] . ' ' . $data['TxnAmount'];
  $binary_output = false;

  // calculate the signed input hash
  echo "hash data: ${hash_data}";
  $hash_value = hash_hmac('sha256', $hash_data, $secret_key, $binary_output);


  $vtenh_request = $data;

  // request payload with signed for vtenh callback
  $vtenh_request["SignedHash"] = $hash_value;

  print_r($vtenh_request);
  // output
  // Array
  // (
  //     [TransactionId] => REF0361472663
  //     [PaymentTokenId] => 16de81d4-b5ef-ef59-16de-81d4b5efef59
  //     [TxnAmount] => 30
  //     [SenderName] => Test User
  //     [SignedHash] => c5b9be690bde7dc8a0abebb1a45c0850359540a4977aecd4cdf13e15a2edfe79
  // )

File: lib/vpago/acleda_mobile/callback_validator.rb
module Vpago
  module AcledaMobile
    class CallbackValidator
      # {
      #   "TransactionId": "REF0361472663",
      #   "PaymentTokenId": "16de81d4-b5ef-ef59-16de-81d4b5efef59",
      #   "TxnAmount": "30",
      #   "SenderName": "Test User",
      #   "SignedHash": "c5b9be690bde7dc8a0abebb1a45c0850359540a4977aecd4cdf13e15a2edfe79",
      #   }
      def initialize(options)
        @options = options
      end

      def call
        valid?
      end

      def secret_key
        ENV['ACLEDA_MOBILE_SECRET_HASH_KEY']
      end

      def valid?
        return false if secret_key.blank?

        hmac_hash == @options[:SignedHash]
      end

      def hmac_hash
        key = secret_key

        message = "#{@options[:TransactionId]} #{@options[:PaymentTokenId]} #{@options[:TxnAmount]}"
  
        OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new('sha256'), key, message)
      end
    end
  end
end
